src/abstractdataframe/reshape.jl

```diff
@@ -215,16 +218,20 @@ function unstack(df::AbstractDataFrame, rowkey::Int, colkey::Int, value::Int)
     col = similar_missing(df[rowkey], length(levs))
     insert!(payload, 1, copy!(col, levs), _names(df)[rowkey])
 end
-unstack(df::AbstractDataFrame, rowkey, colkey, value) =
+unstack(df::AbstractDataFrame, rowkey::ColumnIndex,
+        colkey::ColumnIndex, value::ColumnIndex) =
     unstack(df, index(df)[rowkey], index(df)[colkey], index(df)[value])
 
 # Version of unstack with just the colkey and value columns provided
-unstack(df::AbstractDataFrame, colkey, value) =
+unstack(df::AbstractDataFrame, colkey::ColumnIndex, value::ColumnIndex) =
     unstack(df, index(df)[colkey], index(df)[value])
 
 function unstack(df::AbstractDataFrame, colkey::Int, value::Int)
     # group on anything not a key or value:
-    g = groupby(df, setdiff(_names(df), _names(df)[[colkey, value]]), sort=true)
+    rowkeys = setdiff(_names(df), _names(df)[[colkey, value]])
+    length(rowkeys) == 0 && throw(ArgumentError("No key column found"))
+    length(rowkeys) == 1 && return unstack(df, rowkeys[1], colkey, value) # better performance
```

Comment:
@bkamins do you have the benchmarks you ran to show this has better performance? Just interested in seeing the difference

Reply:
Thank you for the question. I have profiled it earlier and I thought it gave the benefit. Now - given to report it - actually `groupby` approach is a bit faster.

The cost difference is between line (expensive):
```
i = Int(CategoricalArrays.order(refkeycol.pool)[refkeycol.refs[k]])
```
and line (cheap)
```
i = rowkey[k]
```

I have additionally found a bug in `unstack(df::AbstractDataFrame, rowkey::Int, colkey::Int, value::Int)`:
```
julia> using DataFrames

julia> x = DataFrame(id=[missing,1,2,3], a=1:4, b=1:4)
4×3 DataFrames.DataFrame
│ Row │ id      │ a │ b │
├─────┼─────────┼───┼───┤
│ 1   │ missing │ 1 │ 1 │
│ 2   │ 1       │ 2 │ 2 │
│ 3   │ 2       │ 3 │ 3 │
│ 4   │ 3       │ 4 │ 4 │

julia> y = melt(x, :id)
8×3 DataFrames.DataFrame
│ Row │ variable │ value │ id      │
├─────┼──────────┼───────┼─────────┤
│ 1   │ a        │ 1     │ missing │
│ 2   │ a        │ 2     │ 1       │
│ 3   │ a        │ 3     │ 2       │
│ 4   │ a        │ 4     │ 3       │
│ 5   │ b        │ 1     │ missing │
│ 6   │ b        │ 2     │ 1       │
│ 7   │ b        │ 3     │ 2       │
│ 8   │ b        │ 4     │ 3       │

julia> unstack(y)
ERROR: BoundsError: attempt to access 3-element Array{UInt32,1} at index [0]
Stacktrace:
 [1] getindex(::Array{UInt32,1}, ::UInt32) at .\multidimensional.jl:298
 [2] unstack(::DataFrames.DataFrame, ::Int64, ::Int64, ::Int64) at D:\Software\JULIA_PKG\v0.6\DataFrames\src\abstractdataframe\reshape.jl:205
 [3] unstack(::DataFrames.DataFrame) at D:\Software\JULIA_PKG\v0.6\DataFrames\src\abstractdataframe\reshape.jl:254

julia> unstack(y, :variable, :value)
4×3 DataFrames.DataFrame
│ Row │ id      │ a │ b │
├─────┼─────────┼───┼───┤
│ 1   │ 1       │ 2 │ 2 │
│ 2   │ 2       │ 3 │ 3 │
│ 3   │ 3       │ 4 │ 4 │
│ 4   │ missing │ 1 │ 1 │
```

As @nalimilan recommended - I do not want to push too much into one PR.

So my proposal to you is:
1. merge this PR when you feel it is checked enough
2. I have another PR almost ready that I will put on top of this one that
    * removes current implementation of `unstack(df::AbstractDataFrame, rowkey::Int, colkey::Int, value::Int)`
    * introduces `unstack(df::AbstractDataFrame, rowkey::Vector{Int}, colkey::Int, value::Int)` allowing to index on any columns we want;
    * makes old `unstack(df::AbstractDataFrame, rowkey::Int, colkey::Int, value::Int)` and `unstack(df::AbstractDataFrame, colkey::Int, value::Int)` fall back to the above single function.

OK?

Reply:
+1 for another, but if it's not actually faster better remove the comment and the special case from this PR than having incorrect statements in the history.

Reply:
OK - removed the offending line. I will fix the rest of `unstack` after this is merged.